File: Scala/src/com/yourtion/TinyWeb/Controller.scala
package com.yourtion.TinyWeb

/**
  * Created by Yourtion on 9/5/16.
  */
trait Controller {
  def handleRequest(httpRequest: HttpRequest): HttpResponse
}

class FunctionController(view: View, doRequest: (HttpRequest) =>
  Map[String, List[String]] ) extends Controller {

  def handleRequest(request: HttpRequest): HttpResponse =
    try {
      val model = doRequest(request)
      val responseBody = view.render(model)
      HttpResponse(responseBody, 200)
    } catch {
      case e: ControllerException =>
        HttpResponse("", e.getStatusCode)
      case e: RenderingException =>
        HttpResponse("Exception while rendering.", 500)
      case e: Exception =>
        HttpResponse("", 500)
    }

}

File: Clojure/src/com/yourtion/TinyWeb/controller.clj
(ns com.yourtion.TinyWeb.controller
    (:import (com.yourtion.TinyWeb HttpRequest$Builder RenderingException)))

(defn test-controller [http-request]
      {:name (http-request :body)})

(def test-builder (HttpRequest$Builder/newBuilder))

(def test-http-request (.. test-builder (body "Mike") (path "/say-hello") build))

(defn test-controller-with-map [http-request]
      {:name (http-request :body)})

(defn test-view [model]
      (str "<h1>Hello, " (model :name) "</h1>"))

(defn- render [view model]
       (try
         (view model)
         (catch Exception e (throw (RenderingException. e)))))
